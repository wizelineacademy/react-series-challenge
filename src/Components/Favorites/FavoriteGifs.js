import React from "react";
import { connect } from "react-redux";

import Item from "../Gifs/Item";
import { ContainerStyled } from "../../theme/Container.styled";

const FavoriteGifs = ({ favorites: { items, filtered }, searchFavorite }) => {
  const favoritesArr = Object.keys(items);
  if (!favoritesArr.length) return <p>No hay favoritos</p>;

  if (!filtered)
    return (
      <ContainerStyled>
        {favoritesArr.map(key => <Item key={key} item={items[key]} />)}
      </ContainerStyled>
    );

  if (!filtered.length && searchFavorite)
    return (
      <div>
        No se encontraron gifs en favoritos para <b>{searchFavorite}</b>
      </div>
    );

  return (
    <ContainerStyled>
      {filtered
        .filter(item => favoritesArr.find(fav => fav === item.id))
        .map(item => <Item key={item.id} item={item} />)}
    </ContainerStyled>
  );
};

const mapStateToProps = state => {
  const {
    favorites,
    search: { searchFavorite }
  } = state;
  return {
    favorites,
    searchFavorite
  };
};

export default connect(
  mapStateToProps,
  null
)(FavoriteGifs);
