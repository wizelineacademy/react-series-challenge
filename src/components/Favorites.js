import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { toggleFavoriteGif } from '../actions/favorites';
import GifList from './GifList';
import NavBar from './NavBar';

export class Favorites extends Component {
  constructor(props) {
    super(props);
    this.state = {
      favoritesSearch: '',
    };
  }

  handleGifLikeClick = (gif) => {
    this.props.toggleFavoriteGif(gif);
  }

  componentDidCatch() {
    this.setState({ error: 'Error' });
  }

  render() {
    if (this.state.error) {
      return (<h1> Error. Please refresh page </h1>);
    }

    const { gifs } = this.props;
    const filteredGifs = this.getFilteredGifs(gifs);
    return (
      <Fragment>
        <NavBar onSearch={(search) => this.setState({ favoritesSearch: search })} />
        <div>
          <GifList
            onLikeClick={this.handleGifLikeClick}
            gifs={filteredGifs}
          />
        </div>
      </Fragment>
    );
  }

  getFilteredGifs = (gifs) => {
    const { favoritesSearch } = this.state;
    if (favoritesSearch) {
      const filteredGifs = {};
      for (let key in gifs) {
        const item = gifs[key];
        if (item.title.includes(favoritesSearch)) {
          filteredGifs[item.id] = item;
        }
      }
      return filteredGifs;
    } else {
      return gifs;
    }
  }
}

const mapStateToProps = (state = {}) => {
  const { favorites } = state;
  const { gifs } = favorites;
  return { gifs };
}

const mapDispatchToProps = {
  toggleFavoriteGif,
}

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);