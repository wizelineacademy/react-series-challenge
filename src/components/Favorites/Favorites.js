import React, { Component } from 'react'
import { connect } from 'react-redux'
import contentActions from '../../actions/content'
import Searchbar from '../Searchbar'
import Content from '../Content'

const {
  addToFavContent,
  removeFavContent,
} = contentActions.actions


class Favorites extends Component {

  constructor(props) {
    super(props)
    this.toogleFavorite = this.toogleFavorite.bind(this)
    this.state = {
      searchTerm: ''
    }
  }

  toogleFavorite(item, isFav) {
    if(!isFav){
      this.props.addToFavContent(item)
    }
    else {
      this.props.removeFavContent(item)
    }
  }

  handleChange = (e) => {
    this.setState({searchTerm: e.target.value})
  }

  render() {
    return(
      <div>
        <Searchbar
          handleChange={this.handleChange}
        />
        <Content
          data={this.props.content.favorites}
          onClickFav={this.toogleFavorite}
          favorites={this.props.content.favorites}
          filter={this.state.searchTerm}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    content: state.content
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addToFavContent: (item) => {
      dispatch(addToFavContent(item))
    },
    removeFavContent: (item) => {
      dispatch(removeFavContent(item))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Favorites)

