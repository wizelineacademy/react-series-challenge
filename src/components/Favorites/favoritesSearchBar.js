import { SearchBarStyled } from './favorites.style'
import favorites from '../../actions/favorites'
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux'

const mapDispatchToProps = dispatch => {
    /**
     *
     */
    const { creators: { searchFavorite: onInput } } = favorites
    /**
     *
     */
    return bindActionCreators({ onInput }, dispatch)
}

export default connect(null, mapDispatchToProps)(SearchBarStyled)