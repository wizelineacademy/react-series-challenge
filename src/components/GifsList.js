import React from 'react';
import GifCard from './GifCard';

const GifsList = (props) => {

    const renderListGifCard = () => {
        const { gifs } = props;
        return Object.keys(gifs).map((gifId) => <GifCard key={gifId} gif={gifs[gifId]}/>)
    };

    return (
      <div>
          <h5>Gifs List</h5>
          {renderListGifCard()}
      </div>
    );

};

export default GifsList;
