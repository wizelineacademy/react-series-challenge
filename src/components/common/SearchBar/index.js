import SearchBar from './StyledSearchBar';

export default SearchBar
