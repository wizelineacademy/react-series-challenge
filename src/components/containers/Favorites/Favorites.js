import React, { Component } from 'react';
import { FavoritesContainer } from './Favorites.styled';
import { GiphyList, InputSearch, ErrorWatcher } from '../../presentational';
import { connect } from 'react-redux';
import favoriteGifsActions from '../../../actions/favoriteGifs';
const { creators } = favoriteGifsActions;
const { getFavoriteGifsRequested, addFavoriteGifRequested, removeFavoriteGifRequested } = creators;
class Favorites extends Component {
  componentDidMount () {
    const { getFavoriteGifsRequested } = this.props;
    getFavoriteGifsRequested();
  }

  searchGifs (input) {
    const { getFavoriteGifsRequested } = this.props;
    return setTimeout(() => {
      getFavoriteGifsRequested({query: input.value});
    }, 250);
  }

  render() {
    const {
      favoriteGifs,
      getFavoriteGifsRequested,
      addFavoriteGifRequested,
      removeFavoriteGifRequested
    } = this.props;
    return (
      <FavoritesContainer>
        <InputSearch 
          defaultValue={favoriteGifs.query}
          placeholder='Search imagesin my favorites ...' 
          onChange={({target: input}) => {
              if (this.writtingTimmer !== null) {
                clearTimeout(this.writtingTimmer);
                this.writtingTimmer = null;
              }
              this.writtingTimmer = this.searchGifs(input);
          }} />
        <GiphyList 
          dataSource={favoriteGifs}
          onClickFirstPage={() => getFavoriteGifsRequested({offset:1})}
          onClickPreviousPage={() => getFavoriteGifsRequested({movePrevious:true})}
          onClickNextPage={() => getFavoriteGifsRequested({moveNext:true})}
          onClickLastPage={() => getFavoriteGifsRequested({moveLast:true})}
          clickAdd2Favorites={(source) => {
            if (source.isFavorite) {
              removeFavoriteGifRequested(source.id);
            } else {
              addFavoriteGifRequested(source);
            }
          }} />
        <ErrorWatcher visible={favoriteGifs.error !== ''}>{favoriteGifs.error}</ErrorWatcher>
      </FavoritesContainer>
    );
  }
}

const mapStateToProps = (state) => {
  const {favoriteGifs} = state;

  return {
    favoriteGifs
  };
};

export default 
  connect(
    mapStateToProps,
    {
      getFavoriteGifsRequested,
      addFavoriteGifRequested,
      removeFavoriteGifRequested
    }
  )(Favorites);
