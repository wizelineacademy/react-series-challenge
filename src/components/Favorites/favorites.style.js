import styled from 'styled-components'
import gif from '../Gif'
import SearchBar from '../SearchBar'

export const Title = styled.h1`
    font-size: 2rem;
    font-family: sans-serif;
    text-align: center;
`

export const Gifs = styled.div`
    display: grid;
    grid-gap: 0.5rem;
    padding: 2rem 0;
    justify-content: center;
    grid-template-columns: repeat(auto-fill, 12rem);
    grid-template-rows: repeat(auto-fill, 12rem);
`

export const Gif = styled(gif)`
    width: 12rem;
    height: 12rem;
    clip-path: inset(0 round 5%);
    &.favorite {
        clip-path: polygon(25% 0%, 100% 0, 100% 100%, 0 100%, 0% 25%);
    }
`
export const SearchBarStyled = styled(SearchBar)`
    width: 100%;
    padding: 0.5rem 0 0.5rem 1rem;
    background-color: white;
    border: none;
    border-bottom: #858ECC solid thin;
    transition: all 500ms;

    &:focus::placeholder {
        color: #858ECC;
        font-weight: bold;
    }
`

export const FavoritesStyled = styled.section`
    display: flex;
    flex-direction: column;
    padding: 0 2rem;
`