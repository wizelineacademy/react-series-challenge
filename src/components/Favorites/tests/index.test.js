import React from 'react';
import { shallow } from 'enzyme';
import Favorites from '../index';
import GifsList from '../../GifsList';
import SearchBar from '../../SearchBar';
import { Header } from '../../Header';

describe('Favorites component', () => {
  it('should render correctly', () => {
    const component = shallow(<Favorites onSearch={jest.fn()} gifs={[]} />);
    expect(component).toMatchSnapshot();
  });

  it('should have a Header inside', () => {
    const component = shallow(<Favorites onSearch={jest.fn()} gifs={[]} />);
    expect(component.find(Header).length).toBe(1);
  });

  it('should have a SearchBar inside', () => {
    const component = shallow(<Favorites onSearch={jest.fn()} gifs={[]} />);
    expect(component.find(SearchBar).length).toBe(1);
  });

  it('should have a GifsList inside', () => {
    const component = shallow(<Favorites onSearch={jest.fn()} gifs={[]} />);
    expect(component.find(GifsList).length).toBe(1);
  });
});
