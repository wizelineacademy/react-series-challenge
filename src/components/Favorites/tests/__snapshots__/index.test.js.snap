// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    gifs={Array []}
    onSearch={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(GifModal) />,
        <styled.div>
          <styled.h1>
                    Mis favoritos
          </styled.h1>
</styled.div>,
        <SearchBar
          onSearch={undefined}
/>,
        <GifsList
          gifs={Array []}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.h1>
            Mis favoritos
</styled.h1>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Mis favoritos",
          },
          "ref": null,
          "rendered": "Mis favoritos",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSearch": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "gifs": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(GifModal) />,
          <styled.div>
            <styled.h1>
                        Mis favoritos
            </styled.h1>
</styled.div>,
          <SearchBar
            onSearch={undefined}
/>,
          <GifsList
            gifs={Array []}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.h1>
              Mis favoritos
</styled.h1>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Mis favoritos",
            },
            "ref": null,
            "rendered": "Mis favoritos",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSearch": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "gifs": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
